### PR TITLE
db: config: make consistent_cluster_management mandatory

### PR DESCRIPTION
**Pull request by patjed41**
We make `consistent_cluster_management` mandatory in 5.5. This
option will be always unused and assumed to be true.

Additionally, we make `override_decommission` deprecated, as this option
has been supported only with `consistent_cluster_management=false`.

Making `consistent_cluster_management` mandatory also simplifies
the code. Branches that execute only with
`consistent_cluster_management` disabled are removed.

We also update documentation by removing information irrelevant in 5.5.

Fixes scylladb/scylladb#15854

Note about upgrades: this PR does not introduce any more limitations
to the upgrade procedure than there are already. As in
scylladb/scylladb#16254, we can upgrade from the first version of Scylla
that supports the schema commitlog feature, i.e. from 5.1 (or
corresponding Enterprise release) or later. Assuming this PR ends up in
5.5, the documented upgrade support is from 5.4. For corresponding
Enterprise release, it's from 2023.x (based on 5.2), so all requirements
are met.

### CODE REVIEW

**Comment by kbr-scylla** [comment]
> in 6.0

s/6.0/next release/

it appears we plan to release 5.5, which may have this PR

**Comment by kbr-scylla** [comment]
Accompanied by https://github.com/scylladb/scylla-dtest/pull/3770 and https://github.com/scylladb/scylla-pkg/pull/3722

**Comment by kbr-scylla** [comment]
@gleb-cloudius adding you to reviewers just to add some heat to our discussion

File: service/raft/raft_group0.cc
Comment (kbr-scylla) [inline service/raft/raft_group0.cc:1697]: Hm, this is a backwards incompatible **user-visible** change, we should probably keep `2 - aborted` and say that `1` is unused or sth

File: service/raft/raft_group0.hh
Comment (kbr-scylla) [inline service/raft/raft_group0.hh:118]: As above

File: test/boost/group0_test.cc
Comment (kbr-scylla) [inline test/boost/group0_test.cc:67]: This test is in fact evidence why the status numbers must not be modified.

File: docs/architecture/raft.rst
Comment (kbr-scylla) [inline docs/architecture/raft.rst:48]: Hm, that's the question, how we should write this -- if this PR gets into 5.5
cc @annastuchlik -- please review the last commit (` docs: update after making consistent_cluster_management mandatory `)

File: docs/architecture/raft.rst
Comment (kbr-scylla) [inline docs/architecture/raft.rst:52]: what was wrong with "and later"?

File: docs/architecture/raft.rst
Comment (kbr-scylla) [inline docs/architecture/raft.rst:50]: When we add this new upgrade doc, we'll need to include a big note saying "Raft shall be enforced upon you with no way out. Prepare for quorum requirement" or sth (cc @kostja @gleb-cloudius )

File: docs/architecture/raft.rst
Comment (kbr-scylla) [inline docs/architecture/raft.rst:73]: Maybe we should refrain from referring to non-existing (but predicted) versions in documentation for now?
@annastuchlik do you have some standard process for such issues?

File: docs/architecture/raft.rst
Comment (kbr-scylla) [reply docs/architecture/raft.rst:50]: Hmm... @annastuchlik @tzach what do you think of the following idea:

Let's have a "template" upgrade guide present at all times in master branch: "Upgrade guide from <last-release> to NEW_RELEASE_PLACEHOLDER". It would be hidden from users, but visible to developers, so we can update it as we create new PRs. We can also refer to it in other docs; the paragraphs or sentences that refer to it would be hidden. Or maybe NEW_RELEASE_PLACEHOLDER would be automatically replaced with "upcoming release" or something that makes sense.

Once new release is branched, we would have an automatic process that copies this template guide and replaces NEW_RELEASE_PLACEHOLDER with the actual version.

**Comment by kbr-scylla** [comment]
> https://github.com/scylladb/scylladb/pull/16254 was merged with 6.0 in the description...

No worries, this description is for developers, I'm only worried about making references in user-facing docs.

File: test/boost/group0_test.cc
Comment (avikivity) [reply test/boost/group0_test.cc:67]: aye

File: test/alternator/test_table.py
Comment (nyh) [inline test/alternator/test_table.py:468]: I'm not sure we should remove this test-skipping fixture (and its uses). Keeping it for a while longer will allow us to run modern Alternator tests on *older* Scylla. Which is a worthy goal. There is no rush to remove this code now.

File: test/cql-pytest/conftest.py
Comment (nyh) [inline test/cql-pytest/conftest.py:145]: Same here, I don't think there's a rush to remove these functions or their use. Let's allow cql-pytest to run against older Scyllas for a while longer.

File: docs/architecture/raft.rst
Comment (kbr-scylla) [reply docs/architecture/raft.rst:73]: Or leave a reference to the "Verifying" section on the Raft page.
Which there already is: https://opensource.docs.scylladb.com/master/upgrade/upgrade-opensource/upgrade-guide-from-5.1-to-5.2/upgrade-guide-from-5.1-to-5.2-generic.html#validate-raft-setup
But I see it's missing in the ->5.4 upgrade guide.

File: docs/architecture/raft.rst
Comment (patjed41) [reply docs/architecture/raft.rst:52]: My (maybe incorrect) logic was that this section describes enabling Raft without a version upgrade, and that isn't possible in versions later than 5.4. If you have a properly working cluster in 5.5 or later, then you must have already enabled Raft (during rolling upgrade to 5.5 at the latest). So, I thought that this "and later" might be confusing.

But since we are going to remove the "Enabling Raft" section (the discussion below), it doesn't matter anymore.

File: service/raft/raft_group0.cc
Comment (kbr-scylla) [reply service/raft/raft_group0.cc:1697]: Yes

File: docs/architecture/raft.rst
Comment (kbr-scylla) [reply docs/architecture/raft.rst:52]: All versions starting from 5.2 have the "Raft upgrade procedure".

It's true that eventually, this procedure will not be needed (if we forget about recovery for a while) because the version you're upgrading *from* already went through the procedure.

But 5.5 still needs the procedure, because you can have a cluster in 5.4 with Raft disabled, then you do a rolling upgrade. As soon as all nodes have upgraded to version 5.5, that's when the Raft upgrade procedure starts. It starts *after* rolling upgrade, not in the middle of it. For versions which still have the consistent_cluster_management flag -- the procedure starts *after* all nodes enable the flag.

File: test/alternator/test_table.py
Comment (patjed41) [reply test/alternator/test_table.py:468]: From `test/alternator/README.md`:

`run` automatically picks the most-recently compiled version of Scylla in
`build/*/scylla` - but this choice of Scylla executable can be overridden with
the `SCYLLA` environment variable.

I didn't know we can do that. I'll bring these fixtures back.

Comment (patjed41) [reply test/alternator/test_table.py:468]: I also removed the support for the `--raft` option from `test/alternator/run` and `test/cql-pytest/run`. Is it fine? I think this option has already been outdated since `consistent_cluster_management=true` became the default. Tests use raft even without this option unless we specify `consistent_cluster_management=false` directly.

Also, I don't think we want to keep this option for even older Scylla versions with `consistent_cluster_management=false` as the default. Having a flag that normally has no effect seems to be very confusing.

File: docs/architecture/raft.rst
Comment (avikivity) [reply docs/architecture/raft.rst]: Just to clarify, from the previous version to *this* version, where *this* version is the one being documented. So 6.0 documentation should explain how to document to 6.0.

File: service/raft/raft_group0.cc
Comment (patjed41) [reply service/raft/raft_group0.cc]: I think you are right. Also, this `unused` in the description may be understood as if it described the status of the group 0. Group 0 cannot be "unused", so it is misinformation.

**Comment by gleb-cloudius** [comment]
> We make consistent_cluster_management mandatory in 5.5. This option will be always unused and assumed to be true.

When we introduced the option we decided that the option will enable every non experimental cluster management feature that works over raft. Hence the name of the option. If we will make is always enabled it means that all the consistent cluster management related features exiting experimental status will automatically be enabled.

File: docs/architecture/raft.rst
Comment (patjed41) [reply docs/architecture/raft.rst:33]: In that case, I think we should leave the "Enabling Raft" section and keep "Verifying that the Raft upgrade procedure finished successfully" and "Verifying that Raft is enabled" as its subsections. "Enabling Raft" would contain:
- the definition of the Raft upgrade procedure,
- the information that in the current version (5.5) if Raft hasn't already been enabled, it can only be enabled by performing a rolling upgrade from 5.4 and:
  - it will happen automatically,
  - there is no way back.

I don't think we need to keep information about the preconditions like verifying the schema synchronization. They are mentioned in upgrade guides to 5.2 and 5.4 (after merging #16347), and will be mentioned in upgrade guides to 5.5. IIUC, "Enabling Raft" contained these preconditions because they were also necessary when performing the Raft upgrade procedure without the rolling upgrade. In 5.5, the only case, in which the Raft upgrade procedure is performed, is after the rolling upgrade from 5.4. Since we decided to remove all documentation irrelevant in 5.5, I think we can also remove these preconditions from "Raft upgrade procedure".

@kbr-scylla @annastuchlik What do you think about these ideas?

**Comment by kbr-scylla** [comment]
@patjed41 I queued https://github.com/scylladb/scylladb/pull/16350, it should solve at least one reference problem (when you rebase after gating passes)

File: docs/architecture/raft.rst
Comment (kbr-scylla) [reply docs/architecture/raft.rst:33]: Sounds good

File: db/system_keyspace.cc
Comment (patjed41) [reply db/system_keyspace.cc:1878]: Done.

**Comment by patjed41** [comment]
v4: fix for the CI failures above (the new second commit)

**Review by kbr-scylla — APPROVED** [review]

**Comment by avikivity** [comment]
Please add upgrade information to the cover letter. Is there a change in what versions we can upgrade from? If so, what's the change?

**Comment by kbr-scylla** [comment]
IIUC the limitations are no larger than in https://github.com/scylladb/scylladb/pull/16254, i.e. upgrade is possible from the version which first introduced schema commitlog feature (or later). So from 5.1 (or corresponding Enterprise version) or later.

For OSS we only need to support upgrade from 5.4 and for Enterprise only from 2023.x (based on 5.2), so we're good with a spare.

**Review by piodul — APPROVED** [review]

**Comment by kbr-scylla** [comment]
I added this note to the cover letter:

> Note about upgrades: this PR does not introduce any more limitations to the upgrade procedure than there are already. As in scylladb/scylladb#16254, we can upgrade from the first version of Scylla that supports the schema commitlog feature, i.e. from 5.1 (or corresponding Enterprise release) or later. Assuming this PR ends up in 5.5, the documented upgrade support is from 5.4. For corresponding Enterprise release, it's from 2023.x (based on 5.2), so all requirements are met.

**Comment by kbr-scylla** [comment]
@patjed41 according to our discussion with @ptrsmrn in https://github.com/scylladb/scylladb/issues/16338, we should use the new `Deprecated` instead of `Unused` for these config options.

**Comment by patjed41** [comment]
> @patjed41 according to our discussion with @ptrsmrn in #16338, we should use the new `Deprecated` instead of `Unused` for these config options.

Do you mean all these 3: `force_schema_commit_log`, `overwrite_decommission`, `consistent_cluster_management`? `force_schema_commit_log` has already been merged but I think I can change it in this PR.

**Comment by kbr-scylla** [comment]
Yeah, let's do it

File: test/alternator/test_table.py
Comment (nyh) [reply test/alternator/test_table.py]: Yes, I think what you did now is good. The test themselves can still run against ancient Scylla versions if you run it yourself - but the run script only knows how to run a modern version of Scylla - I think that's fine.

If we had any possiblity of turning off raft, we could have left an "--no-raft" option to the run script, but if I understand correctly master no longer allows turning raft off, so we don't even need that. 

So I reviewed your patches and everything seems ok as far as I concerned. But the bigger Raft experts should review it for actual raft content (and documentation changes, etc.) so I didn't leave a review decision. I'll just resolve this discussion. Thanks.

**Comment by patjed41** [comment]
v5: changed `force_schema_commit_log`, `override_decommission`, `consistent_cluster_management` from unused to deprecated (the first change is in the new first commit as it is unrelated)

**Comment by patjed41** [comment]
I think this diff got lost after I resent only a typo fix for the commit message:
https://github.com/scylladb/scylladb/compare/34a7d7254bdaed11186765bb505b72739b1a5182..89d4927482cf51b1a3258d0f100559009616918d

**Comment by patjed41** [comment]
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5256/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5256/) ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6656/)
> #### Failed Tests (1/23442):
> 
>     * [test_disablebinary_and_disablegossip](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5256/testReport/junit/nodetool_additional_test/TestNodetool/Tests___dtest___test_disablebinary_and_disablegossip) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_disablebinary_and_disablegossip)
> 
> 
> #### Build Details:
> 
>     * Duration: 2 hr 27 min
> 
>     * Builder: spider1.cloudius-systems.com

Known flaky test -- #16219. I restarted CI.

**Review by kbr-scylla — APPROVED** [review]

**Pull request merged**
